src/display.rs

```diff
@@ -1,55 +1,213 @@
+                quote!((#fields))
+            }
+            Fields::Named(fields) => {
+                let fields: TokenStream = fields
+                    .named
+                    .iter()
+                    .map(|f| {
+                        let i = f.ident.as_ref().unwrap();
+                        quote!(#i,)
+                    }).collect();
+                quote!({#fields})
+            }
+        }
+    }
+    fn find_meta(&self, attrs: &[Attribute]) -> Result<Option<Meta>> {
+        let mut it = attrs
+            .iter()
+            .filter_map(|a| a.interpret_meta())
+            .filter(|m| m.name() == self.trait_attr);
 
-fn struct_from_str<'a>(trait_name: &str, fields: &[&'a Field]) -> (TokenStream, &'a Type) {
-    if fields.len() != 1 {
-        panic_one_field(trait_name)
-    };
-    let field = &fields[0];
-    let field_ident = &field.ident;
-    let field_type = &field.ty;
-    (quote!(self.#field_ident), field_type)
+        let meta = it.next();
+        if it.next().is_some() {
+            Err(Error::new(meta.span(), "Too many formats given"))
+        } else {
+            Ok(meta)
+        }
+    }
+    fn get_meta_fmt(&self, meta: Meta) -> Result<TokenStream> {
+        let list = match &meta {
+            Meta::List(list) => list,
+            _ => return Err(Error::new(meta.span(), self.get_proper_syntax())),
+        };
+
+        let fmt = match &list.nested[0] {
+            NestedMeta::Meta(Meta::NameValue(MetaNameValue {
+                ident,
+                lit: Lit::Str(s),
+                ..
+            }))
+                if ident == "fmt" =>
+            {
+                s
+            }
+            _ => return Err(Error::new(list.nested[0].span(), self.get_proper_syntax())),
+        };
+
+        let args = list
+            .nested
+            .iter()
+            .skip(1) // skip fmt = "..."
+            .try_fold(TokenStream::new(), |args, arg| {
+                let arg = match arg {
+                    NestedMeta::Literal(Lit::Str(s)) => s,
+                    NestedMeta::Meta(Meta::Word(i)) => {
+                        return Ok(quote_spanned!(list.span()=> #args #i,))
+                    }
+                    _ => return Err(Error::new(arg.span(), self.get_proper_syntax())),
+                };
+                let arg: TokenStream = match arg.parse() {
+                    Ok(arg) => arg,
+                    Err(e) => return Err(Error::new(arg.span(), e)),
+                };
+                Ok(quote_spanned!(list.span()=> #args #arg,))
+            })?;
+
+        Ok(quote_spanned!(meta.span()=> write!(_derive_more_Display_formatter, #fmt, #args)))
+    }
+    fn infer_fmt(&self, span: Span, fields: &Fields) -> Result<TokenStream> {
+        let fields = match fields {
+            Fields::Unit => {
+                return Err(Error::new(
+                    span,
+                    "Can not automatically infer format from unit type",
```

Comment:
Can you instead return the name of the Field here. I got a PR for that functionality in #56 and I think it's indeed quite reasonable, especially now that you can override it it with the `fmt="something"` that this PR adds.